Sources/Commands/SwiftTestTool.swift

```diff
@@ -353,6 +332,181 @@ public struct SwiftTestTool: SwiftTool {
+        ///     - xctestArg: Arguments to pass to XCTest.
+        init(path: AbsolutePath, xctestArg: String? = nil) {
+            self.path = path
+            self.xctestArg = xctestArg
+        }
+
+        /// Constructs arguments to execute XCTest.
+        private func args() -> [String] {
+            var args: [String] = []
+          #if os(macOS)
+            args = ["xcrun", "xctest"]
+            if let xctestArg = xctestArg {
+                args += ["-XCTest", xctestArg]
+            }
+            args += [path.asString]
+          #else
+            args += [path.asString]
+            if let xctestArg = xctestArg {
+                args += [xctestArg]
+            }
+          #endif
+            return args
+        }
+
+        /// Current inherited enviornment variables.
+        private var environment: [String: String] {
+          #if os(Linux)
+            return ProcessInfo.processInfo().environment
+          #else
+            return ProcessInfo.processInfo.environment
+          #endif
+        }
+
+        /// Executes the tests without printing anything on standard streams.
+        ///
+        /// - Returns: A tuple with first bool member indicating if test execution returned code 0
+        ///            and second argument containing the output of the execution.
+        func test() -> (Bool, String) {
+            var output = ""
+            var success = true
+            do {
+                try popen(args(), redirectStandardError: true, environment: environment) { line in
+                    output += line
+                }
+            } catch {
+                success = false
+            }
+            return (success, output)
+        }
+
+        /// Executes and returns execution status. Prints test output on standard streams.
+        func test() -> Bool {
+            let result: Void? = try? system(args(), environment: environment)
+            return result != nil
+        }
+    }
+
+    /// A class to run tests in parallel.
+    final class ParallelTestRunner {
+        /// Path to XCTest binary.
+        private let testPath: AbsolutePath
+
+        /// The producer queue containing list of tests to run.
+        private let producer = SynchronizedQueue<String?>()
+
+        /// The consumer queue containing output of failed tests.
+        private let consumer = SynchronizedQueue<String?>()
+
+        /// Number of parallel workers to spawn.
+        private var numJobs: Int {
+      #if os(Linux)
+        return ProcessInfo.processInfo().activeProcessorCount
+      #else
+        return ProcessInfo.processInfo.activeProcessorCount
+      #endif
+        }
+
+        /// Instance of progress bar. Animating progress bar if stream is a terminal otherwise
+        /// a simple progress bar.
+        private let progressBar: ProgressBarProtocol
+
+        /// A lock to update progress bar status.
+        private var progressBarLock = Basic.Lock()
```

Comment:
(in this case, `run()` would probably be the consumer rather than a separate thread)